peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+		if doneChan != nil {
+			go func() {
+				doneChan <- struct{}{}
+			}()
+		}
+		return
+	}
+	p.outputQueue <- outMsg{msg: msg, doneChan: doneChan}
+}
+
+// QueueInventory adds the passed inventory to the inventory send queue which
+// might not be sent right away, rather it is trickled to the peer in batches.
+// Inventory that the peer is already known to have is ignored.  It is safe for
+// concurrent access.
+func (p *Peer) QueueInventory(invVect *wire.InvVect) {
+	// Don't add the inventory to the send queue if the peer is
+	// already known to have it.
+	if p.isKnownInventory(invVect) {
+		return
+	}
+
+	// Avoid risk of deadlock if goroutine already exited. The goroutine
+	// we will be sending to hangs around until it knows for a fact that
+	// it is marked as disconnected. *then* it drains the channels.
+	if !p.Connected() {
+		return
+	}
+
+	p.outputInvChan <- invVect
+}
+
+// Connected returns whether or not the peer is currently connected.
+func (p *Peer) Connected() bool {
+	return atomic.LoadInt32(&p.connected) != 0 &&
+		atomic.LoadInt32(&p.disconnect) == 0
+}
+
+// Disconnect disconnects the peer by closing the connection.  It also sets
+// a flag so the impending shutdown can be detected.
+func (p *Peer) Disconnect() {
+	// did we win the race?
+	if atomic.AddInt32(&p.disconnect, 1) != 1 {
+		return
+	}
+
+	// Update the address' last seen time if the peer has acknowledged
+	// our version and has sent us its version as well.
+	p.StatsMtx.Lock()
+	if p.verAckReceived && p.versionKnown && p.na != nil {
+		p.addrManager.Connected(p.na)
+	}
+	p.StatsMtx.Unlock()
+
+	log.Tracef("disconnecting %s", p)
+	close(p.quit)
+	if atomic.LoadInt32(&p.connected) != 0 {
+		p.conn.Close()
+	}
+}
+
+// Start begins processing input and output messages.  It also sends the initial
+// version message for outbound connections to start the negotiation process.
+func (p *Peer) Start() error {
+
+	// Already started?
+	if atomic.AddInt32(&p.started, 1) != 1 {
+		return nil
+	}
+
+	log.Tracef("Starting peer %s", p)
+
+	// Send an initial version message if this is an outbound connection.
+	if !p.inbound {
+		err := p.pushVersionMsg()
+		if err != nil {
+			p.LogError("Can't send outbound version message %v", err)
+			p.Disconnect()
+			return err
+		}
+	}
+
+	// Start processing input and output.
+	go p.inHandler()
+	// queueWg is kept so that outHandler knows when the queue has exited so
+	// it can drain correctly.
+	p.queueWg.Add(1)
+	go p.queueHandler()
+	go p.outHandler()
+
+	return nil
+}
+
+// Shutdown gracefully shuts down the peer by disconnecting it.
+func (p *Peer) Shutdown() {
+	log.Tracef("Shutdown peer %s", p)
+	p.Disconnect()
+}
+
+// newPeerBase returns a new base bitcoin peer for the provided server and
+// inbound flag.  This is used by the NewInboundPeer and NewOutboundPeer
+// functions to perform base setup needed by both types of peers.
+func newPeerBase(NewestSha ShaFunc, nonce uint64, donePeers chan *Peer,
```

Comment:
NewestSha shouldn't be capitalized as that goes against effective Go function variable style guidelines.  The same is true for the other cases.